pkg/controllers/node/suite_test.go

```diff
@@ -53,57 +55,117 @@ var _ = AfterSuite(func() {
 })
 
 var _ = Describe("Controller", func() {
+	var provisioner *v1alpha3.Provisioner
+	BeforeEach(func() {
+		provisioner = &v1alpha3.Provisioner{
+			ObjectMeta: metav1.ObjectMeta{Name: v1alpha3.DefaultProvisioner.Name},
+			Spec:       v1alpha3.ProvisionerSpec{},
+		}
+	})
+
 	AfterEach(func() {
 		ExpectCleanedUp(env.Client)
 	})
 
+	Context("Expiration", func() {
+		It("should ignore nodes without TTLSecondsUntilExpired", func() {
+			node := test.Node(test.NodeOptions{
+				Finalizers: []string{v1alpha3.TerminationFinalizer},
+				Labels: map[string]string{
+					v1alpha3.ProvisionerNameLabelKey: provisioner.Name,
+				},
+			})
+			ExpectCreated(env.Client, provisioner, node)
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(node))
+
+			node = ExpectNodeExists(env.Client, node.Name)
+			Expect(node.DeletionTimestamp.IsZero()).To(BeTrue())
+		})
+		It("should ignore nodes without a provisioner", func() {
+			node := test.Node(test.NodeOptions{Finalizers: []string{v1alpha3.TerminationFinalizer}})
+			ExpectCreated(env.Client, provisioner, node)
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(node))
+
+			node = ExpectNodeExists(env.Client, node.Name)
+			Expect(node.DeletionTimestamp.IsZero()).To(BeTrue())
+		})
+		It("should not terminate nodes before expiry", func() {
+			provisioner.Spec.TTLSecondsUntilExpired = ptr.Int64(30)
+			node := test.Node(test.NodeOptions{
+				Finalizers: []string{v1alpha3.TerminationFinalizer},
+				Labels: map[string]string{
+					v1alpha3.ProvisionerNameLabelKey: provisioner.Name,
+				},
+			})
+			ExpectCreated(env.Client, provisioner, node)
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(node))
+
+			node = ExpectNodeExists(env.Client, node.Name)
+			Expect(node.DeletionTimestamp.IsZero()).To(BeTrue())
+		})
+		It("should terminate nodes after expiry", func() {
```

Comment:
Do we have the ability to simulate time going by in these tests? I worry a bit that there might be some code paths not quite exercised here ... compared to having a TTL set (to a non-zero value like 30), prove that (say) 15 seconds in the node is not marked for termination, then (say) 30+ seconds in it actually is so marked?

Reply:
These tests are just copied over since it's a refactor, but I'll look into mocking time for these tests.

Reply:
done